src/server/generic_family.cc

```diff
@@ -463,6 +458,164 @@ void GenericFamily::Stick(CmdArgList args, ConnectionContext* cntx) {
   (*cntx)->SendLong(match_cnt);
 }
 
+struct SortEntryEmpty {};
+
+struct SortEntryScore {
+  double score;
+};
+
+template <
+    // Whether alphanumeric sort will be used.
+    bool ALPHA,
+    // Deduced StorageKind type for optimal size.
+    container_utils::StorageKind STORAGE_KIND,
+    // Whether to store double score. It will be stored for non-alphanumeric sorts
+    // when our items are not trivailly convertiable to doubles (i.e. they are not integers)
+    bool STORE_SCORE = !ALPHA && STORAGE_KIND != container_utils::StorageKind::INTEGRAL>
+struct SortEntry : public std::conditional_t<STORE_SCORE, SortEntryScore, SortEntryEmpty> {
+  // Key type. If we sort alphanumerically, we store strings. Else we mimic
+  // the container storage type.
+  using KEY_TYPE =
+      std::conditional_t<ALPHA, std::string, container_utils::StorageType<STORAGE_KIND> >;
+  // Comparison type. std:string for alphanumeric sort, double otherwise.
+  using COMPARE_TYPE =
+      std::conditional_t<ALPHA, const std::string&,
+                         // If we store only integers, we can use them for comparison.
+                         std::conditional_t<std::is_integral_v<KEY_TYPE>, KEY_TYPE, double> >;
+
+  // This is either: std::string, int64_t or std::variant<std::string, int64_t>
+  KEY_TYPE key;
+
+  bool Parse(std::string&& item) {
+    if constexpr (STORE_SCORE) {
+      if (!absl::SimpleAtod(item, &this->score))
+        return false;
+    }
+    if constexpr (std::is_assignable_v<KEY_TYPE&, std::string&&>) {
+      key = std::move(item);
+    } else {
+      assert(false && "unreachable");
```

Comment:
can it be `static_assert` ?

Reply:
No, because all code paths are compiled. The `STORAGE_KIND` is just a kind of "agreement" that does not exist on the type level.
It could, however, be part of the type system if we wire it down to the Iterate functions.  

I don't like this part either, but I'm not sure the alternative is worth the effort